### PR TITLE
Optimize rich text rewrite handlers via bulk lookups

### PR DESCRIPTION
**Pull request by chosak**
Rich text rewrite handlers currently work on each link/embed tag independently: first the type of tag is determined (for example, a page link), then the related model instance is retrieved from the database, and then the instance is used to generate the HTML representation. This creates some inefficiency: for example, a rich text field with 10 page links requires 20 database queries (one each for the base page model and one each for the specific page model).

This PR modifies the rich text rewrite logic so that it instead groups and processes all link/embed tags of the same type at once. This allows us to optimize the database lookups by making significantly fewer queries, for example retrieving all images and documents in a single call and retrieving all pages in the minimal number of calls by leveraging `PageQuerySet.specific()`. For the example above, a rich text field with 10 page links of the same page type would now require only 2 database queries.

To accomplish this I expanded the definition of the `EntityHandler` class to add two new methods, `get_many` and `expand_db_attributes_many`. These changes should be backwards compatible with any custom rich text rewrite handlers that are out there. The main interface change is that the (undocumented) `EntityRewriter` class now supports the notion of both individual and bulk rewrite functions -- I played around with a few approaches for this and this seemed the best way to support the dual functionality for both the frontend and editor rich text representations.

### CODE REVIEW

**Comment by lb-** [comment]
Due to no review for some time, just adding needs work label as this will need some rebase / redo on the latest code.

**Comment by lb-** [comment]
Ready for review again.

**Review by zerolab — CHANGES_REQUESTED** [review]
So much goodness here. Thank you @chosak.

Left a number of questions and suggestions

File: wagtail/rich_text/__init__.py [hunk 1]
```diff
@@ -130,6 +131,14 @@ def get_instance(cls, attrs: dict) -> Model:
         model = cls.get_model()
         return model._default_manager.get(id=attrs["id"])
 
+    @classmethod
+    def get_many(cls, attrs_list: List[dict]) -> List[Model]:
+        model = cls.get_model()
+        instance_ids = [attrs.get("id") for attrs in attrs_list]
+        instances_by_id = model._default_manager.in_bulk(instance_ids)
+        instances_by_str_id = {str(k): v for k, v in instances_by_id.items()}
+        return [instances_by_str_id.get(str(id_)) for id_ in instance_ids]
```
[the remark this answers is not in the document]
Comment (chosak) [reply wagtail/rich_text/__init__.py:140]: That's right -- IDs in rich text come as strings, for example from `<a linktype="page" id="123">`. The bulk lookup uses Django's [`in_bulk`](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#in-bulk) which works when using strings as the ID input but returns its results using the actual type of the primary key (which may be integer, or string, or even something else, depending on the model being fetched):

```
>>> pages = Page.objects.in_bulk(["1", "2", "3"])
>>> pages["1"]
Traceback (most recent call last):
  File "<console>", line 1, in <module>
KeyError: '1'
>>> pages[1]
<Page: Root>
```

So we have a list of entity IDs as strings from rich text -- we use these to look up objects in bulk which gives us a key list of IDs of varying types -- we then need to convert these IDs back to strings so we can match them back to what we had in rich text.

[the remark this answers is not in the document]
Comment (zerolab) [reply wagtail/rich_text/__init__.py:140]: I figured this may have been the case. thank you

File: wagtail/rich_text/rewriters.py [hunk 2]
```diff
@@ -25,101 +27,172 @@ def extract_attrs(attr_string):
...
+        except KeyError:
+            rule = None
+
+        if not rule:
+            try:
+                rule = self.rules[tag_type]
+            except KeyError:
+                pass
+            else:
+                rule = self.convert_rule_to_bulk_rule(rule)
```
[the remark this answers is not in the document]
Comment (chosak) [reply wagtail/rich_text/rewriters.py:127]: @zerolab alternatively we could drop support for non-bulk rules. The Wagtail documentation [describes](https://docs.wagtail.org/en/stable/extending/rich_text_internals.html#rewrite-handlers) implementing custom rewrite handlers that inherit from either EmbedRewriter or LinkRewriter; the ability to define some other rewriter class is technically possible but not really supported. The reason I left the non-bulk functionality in was to avoid making a breaking change - what would you think about doing that? I can certainly add tests but it would be testing a code path that wouldn't be used by the Wagtail code any longer.

File: wagtail/tests/test_rich_text.py [hunk 3]
```diff
@@ -101,6 +119,78 @@ def test_expand_db_html_with_embed(self, get_embed):
...
+minimal number of database queries (3) to expand them.
+
+Both of these images should be retrieved with 3 queries, one to fetch the
+image objects in bulk and then one per image to fetch their renditions:
+
+This is an image: <embed embedtype="image" id="1" format="left" />
+This is another image: <embed embedtype="image" id="2" format="left" />
+        """
+            )
+
```
[the remark this answers is not in the document]
Comment (chosak) [reply wagtail/tests/test_rich_text.py:193]: The embed code doesn't currently support bulk fetching of embeds; see for example [embeds.py](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py).

[The current embed tests](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/tests/test_rich_text.py) should properly test the existing rich text conversion (including the parts touched by this PR).

Adding a test specifically to count database queries when doing embed rich text conversion didn't seem valuable to me because that code [does a query](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py#L17) and then [does a save](https://github.com/wagtail/wagtail/blob/e03c41250776612468226597a077ba1fc2d9b865/wagtail/embeds/embeds.py#L62) (which happens in a transaction), so using `assertNumQueries` doesn't feel particularly useful.

**Comment by chosak** [comment]
Thank you for the thorough review @zerolab! I've tried to address most of your comments, see a few open questions above.

File: wagtail/rich_text/rewriters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zerolab) [reply wagtail/rich_text/rewriters.py:129]: You are right in that we do say

> New handlers must be Python classes that inherit from either `wagtail.richtext.LinkHandler` or `wagtail.richtext.EmbedHandler`. Your new classes should override at least some of the following methods (listed here for `LinkHandler`, although `EmbedHandler` has an identical signature)


We have two places in which we invoke the the page and embed rewriters:

1. https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/rich_text/__init__.py#L26 which you've update in this PR.
2. and https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/admin/rich_text/converters/editor_html.py#L146 -- which needs updating

Usually we follow [a deprecation process](https://docs.wagtail.org/en/stable/contributing/release_process.html#deprecation-policy). While not an official API, I could bet folks have customised or monkey-patched things so that if we made the breaking change, we'd hear. @gasman what do you think?

File: wagtail/rich_text/pages.py [hunk 4]
```diff
@@ -12,16 +15,24 @@ def get_model():
...
     @classmethod
-    def expand_db_attributes(cls, attrs):
-        try:
-            page = cls.get_instance(attrs)
-            return '<a href="%s">' % escape(page.localized.specific.url)
-        except Page.DoesNotExist:
-            return "<a>"
+    def expand_db_attributes_many(cls, attrs_list: List[dict]) -> List[str]:
+        return [
+            '<a href="%s">' % escape(page.localized.specific.url) if page else "<a>"
```
[the remark this answers is not in the document]
Comment (zerolab) [reply wagtail/rich_text/pages.py:33]: It is worth pointing out that we have another `PageLinkHandler`, used as a [converter](https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/admin/wagtail_hooks.py#L425) for editor html which also has [`expand_db_attributes`](https://github.com/wagtail/wagtail/blob/46cb06a8604b68594757117fb7794fb0145a5e99/wagtail/admin/rich_text/converters/editor_html.py#L170-L172)

One for a follow up PR, I think

**Review by zerolab — COMMENTED** [review]
This is shaping up sooo nicely 👏🏼 

Left a few more notes, one being prompted by an update on another semi-related PR

File: wagtail/rich_text/pages.py [hunk 5]
```diff
@@ -12,16 +15,24 @@ def get_model():
         return Page
 
     @classmethod
-    def get_instance(cls, attrs):
-        return super().get_instance(attrs).specific
+    def get_many(cls, attrs_list: List[dict]) -> List[Model]:
+        # Override LinkHandler.get_many to reduce database queries through the
+        # use of PageQuerySet.specific() instead of QuerySet.in_bulk().
+        instance_ids = [attrs.get("id") for attrs in attrs_list]
+        qs = Page.objects.filter(id__in=instance_ids).defer_streamfields().specific()
```
[the remark this answers is not in the document]
Comment (chosak) [reply wagtail/rich_text/pages.py]: @zerolab I misunderstood this suggestion at first -- `specific_deferred` can't be applied to a PageQuerySet, can it? It has to be applied page by page.

Is your suggestion to do something like:

```py
qs = [page.specific_deferred for page in Page.objects.filter(id__in=instance_ids)]
```

If I understand right this means that we don't ever fetch the specific tables at all?

**Comment by chosak** [comment]
@zerolab thanks for your continued feedback. I've pushed a commit that makes the invocation of MultiRuleRewriter here a little more explicit:

https://github.com/wagtail/wagtail/blob/9d5d4ade48ca22479809e64c7ee4143b6af1c58a/wagtail/admin/rich_text/converters/editor_html.py#L146-L148

It's a little unfortunate that there's this other identically-named PageLinkHandler class. The editor_html rich text functionality could also be optimized as you observe; this seemed like less of a performance consideration because if I understand this is happening at save time, as opposed to page render time.

File: wagtail/rich_text/pages.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zerolab) [reply wagtail/rich_text/pages.py:22]: uh-oh, there I was a bit overexcited. You're right, that is on the page object, not the queryset. 

We could probably improve this further (by only selecting the fields involved in the URL generation, but that may be premature)

[the remark this answers is not in the document]
Comment (zerolab) [reply wagtail/rich_text/pages.py:22]: Resolving conversation as nothing more to do

File: wagtail/admin/rich_text/converters/editor_html.py [hunk 6]
```diff
@@ -143,7 +143,9 @@ def html_rewriter(self):
             elif isinstance(rule, LinkTypeRule):
                 link_rules[rule.link_type] = rule.handler.expand_db_attributes
 
-        return MultiRuleRewriter([LinkRewriter(link_rules), EmbedRewriter(embed_rules)])
+        return MultiRuleRewriter(
+            [LinkRewriter(rules=link_rules), EmbedRewriter(rules=embed_rules)]
```
Comment (zerolab) [inline wagtail/admin/rich_text/converters/editor_html.py:147]: @chosak should these not be `bulk_rules` a la https://github.com/wagtail/wagtail/pull/5875/files#diff-28e86f41edadc0bbf445852a6284888ab58ce7b2947725fccf60fbd189085698R30? or am I forgetting something?

Comment (chosak) [reply wagtail/admin/rich_text/converters/editor_html.py:147]: It's confusing!

The richtext rewriters LinkRewriter and EmbedRewriter currently (in main, before this PR) have a signature like

```py
LinkRewriter(rules, reference_extractors=None)
```

where `rules` is a mapping of `{linktype: expand_db_attributes}`, for example, given a linktype `page`, call method `PageLinkHandler.expand_db_attributes`. Each method takes a dict of attributes `{"linktype": "page", "id": 1234}` and returns the object being referenced (e.g. the page with ID 1234).

What this PR does is add another bulk approach to doing these lookups, such that the signature now looks like

```py
LinkRewriter(rules=None, bulk_rules=None, reference_extractors=None)
```

If bulk rules are provided, they will be used instead; for example this PR defines a `PageLinkHandler.expand_db_attributes_many` that allows for bulk lookup of page objects given multiple attribute dicts.

This PR reimplements all of the _DB -> frontend HTML_ rewriters so that they have bulk implementations. It's possible that devs have implemented other non-bulk custom rewriters, especially since this interface has been documented.

The _rich text editor -> DB_ rewrite path uses different rewrite rules (named similarly, unfortunately) that that this PR has not rewritten to work in bulk. I could try to migrate these to work in bulk, but I'm not sure the implementation works similarly enough for that to make sense (note for example that [these are `WhitelistRule` instances](https://github.com/wagtail/wagtail/blob/c11db31a787ab68df79bb051838dc51633a8b2ee/wagtail/admin/rich_text/converters/editor_html.py#L124-L127) that do a lot of cleanup that doesn't involve fetching things from the database).

In other words, the two different rewrite paths seemingly use the same framework (MultiRuleRewriter, LinkRewriter, EmbedRewriter) to accomplish different tasks.

Comment (zerolab) [reply wagtail/admin/rich_text/converters/editor_html.py:147]: You're right. Despite my double checking, I muddled this up again. Long day!

Thank you for taking the time to spell it out clearly. May I suggest a developer note just above `MultiRuleRewriter()` here to hint and warn future us of the confusing naming (e.g. "The rich text editor -> DB rewrite rules are different than the DB -> front end ones. The latter accept `bulk_rules` kwarg for bulk lookups"

Comment (zerolab) [reply wagtail/admin/rich_text/converters/editor_html.py:147]: I think aside from this, and the decision on https://github.com/wagtail/wagtail/pull/5875#discussion_r1493739982 on whether we 

a. do nothing and convert rules to bulk rules
b. add a deprecation notice for custom rewriters
c. just drop support for non-bulk rules for obscure rewriters since you handle it for `LinkHandler` and `EmbedHandler`s

Comment (chosak) [reply wagtail/admin/rich_text/converters/editor_html.py:147]: Hi @zerolab, just checking back in on this. How can I help move this PR along?

Comment (zerolab) [reply wagtail/admin/rich_text/converters/editor_html.py:147]: Assigned to @gasman for this last bit.

File: wagtail/rich_text/rewriters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/rich_text/rewriters.py:129]: @chosak @zerolab If I'm following correctly, LinkRewriter / EmbedRewriter (note: NOT LinkHandler / EmbedHandler) are entirely undocumented and internal to Wagtail, and so we should feel at liberty to make breaking changes to it without a formal deprecation process - at most we would add a "changes to undocumented internals" release note, if we have reason to believe it's being used in the wild.

However, it sounds like `wagtail.admin.rich_text.converters.editor_html` _is_ still invoking these classes with non-bulk rules. I'm pretty sure that module only exists to support non-Draftail editors such as [wagtail-hallo](https://github.com/wagtail-nest/wagtail-hallo/), and we're probably way overdue for offloading this code out of Wagtail and into the packages in question (if indeed any are still being maintained) - but that's too much scope creep for this PR. So, I'd suggest keeping the `convert_rule_to_bulk_rule` code path in place for the purposes of keeping the editor_html module limping along, and adding the missing test coverage for editor_html to ensure that this code path is tested. It looks like [`EditorHTMLConverter.from_database_format`](https://github.com/wagtail/wagtail/blob/dd48762b188c355a763eeb2a0d6a4c0b0d9fb752/wagtail/admin/rich_text/converters/editor_html.py#L148-L149) is the entry point in question, and the test would look something like the following (in [TestPageLinkHandler](https://github.com/wagtail/wagtail/blob/dd48762b188c355a763eeb2a0d6a4c0b0d9fb752/wagtail/admin/tests/test_rich_text.py#L490)):

```python
    def test_editorhtmlconverter_from_database_format(self):
        events_page_id = Page.objects.get(url_path="/home/events/").pk
        db_html = '<a linktype="page" id="%d">foo</a>' % events_page_id
        converter = EditorHTMLConverter(features=["link"])
        editor_html = converter.from_database_format(db_html)
        self.assertEqual(
            result,
            '<a data-linktype="page" data-id="%d" data-parent-id="2" href="/events/">foo</a>'
            % events_page_id,
        )
```

File: docs/extending/rich_text_internals.md [hunk 7]
```diff
@@ -106,14 +114,18 @@ You can create custom rewrite handlers to support your own new `linktype` and `e
...
+        Optional. The classmethod ``get_instance`` method also only applies to those handlers that are used to render content related to Django models. This method is expected to take a dictionary of attributes from a database rich text ``<a>`` tag (``<embed>`` for ``EmbedHandler``) and use it to return the specific Django model instance being referred to.
 
         For example, ``PageLinkHandler.get_instance`` might receive ``{'id': 123}`` and return the instance of the Wagtail ``Page`` class with ID 123.
 
         If left undefined, a default implementation of this method will query the ``id`` model field on the class returned by ``get_model`` using the provided ``id`` attribute; this can be overridden in your own handlers should you want to use some other model field.
+
+    .. method:: get_many(attrs_list)
+
+        Optional. The classmethod ``get_many`` method works similarly to ``get_instance`` but instead takes a list of attribute dictionaries and returns a list of Django model instances.
```
Comment (gasman) [inline docs/extending/rich_text_internals.md:125]: I think it's worth documenting what happens when an instance is not found - namely, returning `None` for that item in the list. (And I suppose for completeness the documentation for `get_instance` should mention that it throws an `ObjectDoesNotExist` exception in that situation.)

File: wagtail/rich_text/rewriters.py [hunk 8]
```diff
@@ -25,101 +28,158 @@ def extract_attrs(attr_string):
     return attributes
 
 
-class EmbedRewriter:
-    """
-    Rewrites <embed embedtype="foo" /> tags within rich text into the HTML fragment given by the
```
Comment (gasman) [inline wagtail/rich_text/rewriters.py:30]: Can we reinstate this docstring (and the one for LinkRewriter) please?

**Review by gasman — CHANGES_REQUESTED** [review]

File: docs/extending/rich_text_internals.md (hunk 7 above)
Comment (chosak) [reply docs/extending/rich_text_internals.md:128]: Thanks for this suggestion -- the existing `MediaEmbedHandler` [raises `EmbedUnsupportedProviderException`](https://github.com/wagtail/wagtail/blob/f30c65138e1cb2162134f15976aec4b1b383752c/wagtail/embeds/embeds.py#L32) if the embed URL isn't supported, not `ObjectDoesNotExist`.

I used a slightly less specific wording that tries to capture this case:

> This method should raise an exception if the provided attributes cannot be used to retrieve a Django model instance, for example if the provided `id` attribute is invalid.

File: wagtail/rich_text/rewriters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chosak) [reply wagtail/rich_text/rewriters.py:135]: Great suggestion, thanks. Implemented in ab8ba8c.

**Comment by chosak** [comment]
Thanks for the review @gasman! I've tried to address your suggestions.

**Review by zerolab — APPROVED** [review]
This LGTM!

@gasman there was that [last comment from you](https://github.com/wagtail/wagtail/pull/5875/files#r1519034154) that @chosak addressed in a more generic way

**Review by gasman — APPROVED** [review]
Thanks @chosak and @zerolab! Yep, happy with this too.

**Comment by gasman** [comment]
Merged in e32aee036f
